mobile/examples/object_detection/ios/ORTObjectDetection/ModelHandler.swift

```diff
@@ -0,0 +1,448 @@
+        
+        if numDetections == 0 {
+            return resultsArray
+        }
+        
+        for i in 0 ... numDetections - 1 {
+            let score = detectionScores[i]
+            
+            // Filter results with score < threshold.
+            guard score >= threshold else {
+                continue
+            }
+            
+            let detectionClassIndex = Int(detectionClasses[i])
+            let detectionClass = labels[detectionClassIndex + 1]
+            
+            var rect = CGRect.zero
+            
+            // Translate the detected bounding box to CGRect.
+            rect.origin.y = CGFloat(detectionBoxes[4 * i])
+            rect.origin.x = CGFloat(detectionBoxes[4 * i + 1])
+            rect.size.height = CGFloat(detectionBoxes[4 * i + 2]) - rect.origin.y
+            rect.size.width = CGFloat(detectionBoxes[4 * i + 3]) - rect.origin.x
+            
+            let newRect = rect.applying(CGAffineTransform(scaleX: width, y: height))
+            
+            let colorToAssign = colorForClass(withIndex: detectionClassIndex + 1)
+            let inference = Inference(score: score,
+                                      className: detectionClass,
+                                      rect: newRect,
+                                      displayColor: colorToAssign)
+            resultsArray.append(inference)
+        }
+        
+        // Sort results in descending order of confidence.
+        resultsArray.sort { first, second -> Bool in
+            first.score > second.score
+        }
+        
+        return resultsArray
+    }
+    
+    // This method preprocesses the image by cropping pixel buffer to biggest square and scaling the cropped image to model dimensions.
+    private func preprocess(
+        ofSize size: CGSize,
+        _ buffer: CVPixelBuffer
+    ) -> CVPixelBuffer? {
+        let imageWidth = CVPixelBufferGetWidth(buffer)
+        let imageHeight = CVPixelBufferGetHeight(buffer)
+        let pixelBufferType = CVPixelBufferGetPixelFormatType(buffer)
+        
+        assert(pixelBufferType == kCVPixelFormatType_32BGRA ||
+            pixelBufferType == kCVPixelFormatType_32ARGB)
+        
+        let inputImageRowBytes = CVPixelBufferGetBytesPerRow(buffer)
+        let imageChannels = 4
+        
+        CVPixelBufferLockBaseAddress(buffer, CVPixelBufferLockFlags(rawValue: 0))
+        
+        // Find the biggest square in the pixel buffer and advance rows based on it.
+        guard let inputBaseAddress = CVPixelBufferGetBaseAddress(buffer) else {
+            return nil
+        }
+        
+        // Get vImage_buffer
+        var inputVImageBuffer = vImage_Buffer(
+            data: inputBaseAddress, height: UInt(imageHeight), width: UInt(imageWidth),
+            rowBytes: inputImageRowBytes
+        )
+        
+        let scaledRowBytes = Int(size.width) * imageChannels
+        guard let scaledImageBytes = malloc(Int(size.height) * scaledRowBytes) else {
```

Comment:
>malloc [](http://example.com/codeflow?start=37&length=6)

one idea to easily handle the free - how about having `defer { free(scaledImageBytes) }` and setting scaledImageBytes to nil after handing it over to CVPixelBufferCreateWithBytes()? free(0) (and, I assume, free(nil)) should be a no-op

Reply:
Not so sure how to do the part "set the scaledImageBytes to nil.." because I guess I set scaledImageBytes to be a `let`. But thanks for the suggestion,I understand that we should call free wherever a possible error can cause exit. I added defer{free(scaledImageBytes)} inside`guard scaleError == kvImageNoError`.

Reply:
right - scaledImageBytes would have to be declared with var instead